### PR TITLE
Lightmapper refactor and cleanup

### PR DESCRIPTION
**Pull request by mvaligursky**
- refactoring of Lightmapper class into cleaner code broken down into functions, with larger parts being replaced
- no public API changes
- ref counted lighmaps owned by meshInstances, they get automatically released when no longer referenced
- side effect: fixed memory leaks after each bake in the Editor (and possibly engine)
- side effect: no longer printing errors when setting vec3 parameter to vec4 (related to shaders not being compiled for used light)
- side effect: no longer fails to bake few meshes in my test scene
- example: lights-baked-ao.html - non-public example showing how to bake skydome AO (not very efficient in this implementation, more of a proof of concept at this point)

Screen Shot 2021-02-03 at 3 58 47 PM

### CODE REVIEW

File: examples/graphics/lights-baked-render.html [hunk 1]
```diff
@@ -155,12 +157,21 @@
...
         app.lightmapper.bake(null, bakeType);
+//        app.lightmapper.bake([ground], bakeType);
 
         // Set an update function on the app's update event
         var time = 4;
+        var bake = true;
         app.on("update", function (dt) {
             time += dt;
 
+            if (time > 7 && bake) {
```
Comment (Maksims) [inline examples/graphics/lights-baked-render.html:168]: What this delay is for?

Comment (mvaligursky) [reply examples/graphics/lights-baked-render.html:168]: it's just my test code while refactoring to re-bake one one time to test memory release and stuff. It won't be there when done.

File: src/scene/lightmapper.js [hunk 2]
```diff
@@ -246,337 +495,303 @@ class Lightmapper {
...
+            meshInstances[i].material = this.materials[i];
+        }
+    }
 
-        var renderOrModelMeshInstances = function (node) {
-            return node.render ? node.render.meshInstances : node.model.model.meshInstances;
-        };
+    // preparas camera / frustum of the light for rendering the bakeNode
+    // returns true if light affects the node
+    LightCameraPrepare(bakeLight, bakeNode, shadowCam) {
```
Comment (willeastcott) [inline src/scene/lightmapper.js:682]:
```suggestion
    lightCameraPrepare(bakeLight, bakeNode, shadowCam) {
```

File: src/scene/lightmapper.js [hunk 3]
```diff
@@ -589,122 +804,86 @@ class Lightmapper {
...
-            }
+            for (node = 0; node < bakeNodes.length; node++) {
 
-            for (node = 0; node < nodes.length; node++) {
+                let bakeNode = bakeNodes[node];
+                rcv = bakeNode.meshInstances;
 
-                rcv = nodesMeshInstances[node];
-                bounds = nodeBounds[node];
+                const lightAffectsNode = this.LightCameraPrepare(bakeLight, bakeNode, shadowCam);
```
Comment (willeastcott) [inline src/scene/lightmapper.js:845]:
```suggestion
                const lightAffectsNode = this.lightCameraPrepare(bakeLight, bakeNode, shadowCam);
```

File: src/scene/lightmapper.js (hunk 2 above)
Comment (mvaligursky) [reply src/scene/lightmapper.js:682]: thanks, great catch ..

File: examples/graphics/lights-baked-ao.html [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        var ground = new pc.Entity("Ground");
+        ground.addComponent('render', {
+            castShadows: false,
+            castShadowsLightmap: false,
+            lightmapped: true,
+            type: "plane",
+            material: material
+        });
+        app.root.addChild(ground);
+        ground.setLocalPosition(0, 0, 0);
```
Comment (willeastcott) [inline examples/graphics/lights-baked-ao.html:59]: Redundant - remove.

File: examples/graphics/lights-baked-ao.html [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+                lightmapped: true,
+                type: shape,
+                material: material
+            });
+            app.root.addChild(entity);
+
+            // random orientation
+            var scale = 1 + Math.random() * 2;
+            entity.setLocalScale(scale, scale, scale);
+            entity.setLocalPosition(Math.random() * 30 - 15, 1 * scale, Math.random() * 30 - 15);
```
Comment (willeastcott) [inline examples/graphics/lights-baked-ao.html:95]:
```suggestion
	            entity.setLocalPosition(Math.random() * 30 - 15, scale, Math.random() * 30 - 15);
```

File: src/scene/lightmapper.js [hunk 6]
```diff
@@ -19,104 +19,88 @@ import { drawQuadWithShader } from '../graphics/simple-post-effect.js';
...
-var bounds = new BoundingBox();
-var lightBounds = new BoundingBox();
-var tempSphere = {};
+const PASS_COLOR = 0;
+const PASS_DIR = 1;
 
-var PASS_COLOR = 0;
-var PASS_DIR = 1;
+let tempVec = new Vec3();
+let tempSphere = {};
```
Comment (willeastcott) [inline src/scene/lightmapper.js:44]:
```suggestion
let tempSphere = new pc.BoundingSphere();
```

File: src/scene/lightmapper.js [hunk 7]
```diff
@@ -19,104 +19,88 @@ import { drawQuadWithShader } from '../graphics/simple-post-effect.js';
...
+        // bounds for non-directional light
+        if (light.type !== LIGHTTYPE_DIRECTIONAL) {
+
+            // world sphere
+            light._node.getWorldTransform();
+            light.getBoundingSphere(tempSphere);
+
+            // world aabb
+            this.lightBounds = new BoundingBox();
+            this.lightBounds.center = tempSphere.center;
```
Comment (willeastcott) [inline src/scene/lightmapper.js:87]:
```suggestion
            this.lightBounds.center.copy(tempSphere.center);
```

File: src/scene/lightmapper.js [hunk 8]
```diff
@@ -19,104 +19,88 @@ import { drawQuadWithShader } from '../graphics/simple-post-effect.js';
...
+            this.lightBounds.halfExtents.x = tempSphere.radius;
+            this.lightBounds.halfExtents.y = tempSphere.radius;
+            this.lightBounds.halfExtents.z = tempSphere.radius;
```
Comment (willeastcott) [inline src/scene/lightmapper.js:90]:
```suggestion
            this.lightBounds.halfExtents.set(tempSphere. radius, tempSphere. radius, tempSphere. radius);
```

File: src/scene/lightmapper.js [hunk 9]
```diff
@@ -149,23 +139,250 @@ class Lightmapper {
...
+                type: TEXTURETYPE_RGBM
+            });
+            this.blackTex.name = 'lightmapBlack';
+
+            // incref black texture in the cache to avoid it being destroyed
+            MeshInstance.incRefLightmap(this.blackTex);
+
+            // camera used for baking
+            let camera = new Camera();
+            camera.clearColor = new Color(0, 0, 0, 0);
```
Comment (willeastcott) [inline src/scene/lightmapper.js:185]:
```suggestion
            camera.clearColor.set(0, 0, 0, 0);
```

File: src/scene/lightmapper.js [hunk 10]
```diff
@@ -149,23 +139,250 @@ class Lightmapper {
...
+        for (let pass = 0; pass < passCount; pass++) {
+            if (!this.passMaterials[pass]) {
+
+                let lmMaterial = new StandardMaterial();
+                lmMaterial.chunks.transformVS = xformUv1; // draw UV1
+
+                if (pass === PASS_COLOR) {
+                    lmMaterial.chunks.endPS = bakeLmEnd; // encode to RGBM
+                    // don't bake ambient
+                    lmMaterial.ambient = new Color(0, 0, 0);
```
Comment (willeastcott) [inline src/scene/lightmapper.js:238]:
```suggestion
                    lmMaterial.ambient.set(0, 0, 0);
```

File: src/scene/lightmapper.js [hunk 11]
```diff
@@ -149,23 +139,250 @@ class Lightmapper {
...
+
+    // recursively walk the hierarchy of nodes starting at the specified node
+    // collect all nodes that need to be lightmapped to bakeNodes array
+    // collect all nodes with geometry to allNodes array
+    collectModels(node, bakeNodes, allNodes) {
+        if (!node.enabled) return;
+
+        // mesh instances from model component
+        let meshInstances;
+        if (node.model && node.model.model && node.model.enabled) {
```
Comment (willeastcott) [inline src/scene/lightmapper.js:290]:
```suggestion
        if (node.model?.model?.enabled) {
```

File: src/scene/lightmapper.js [hunk 12]
```diff
@@ -149,23 +139,250 @@ class Lightmapper {
...
+            if (allNodes) allNodes.push(new MeshNode(node));
+            if (node.model.lightmapped) {
+                if (bakeNodes) {
+                    meshInstances = node.model.model.meshInstances;
+                }
+            }
+        }
+
+        // mesh instances from render component
+        if (node.render && node.render.enabled) {
```
Comment (willeastcott) [inline src/scene/lightmapper.js:300]:
```suggestion
        if (node.render?.enabled) {
```

File: src/scene/lightmapper.js [hunk 13]
```diff
@@ -149,23 +139,250 @@ class Lightmapper {
...
+            let hasUv1 = true;
+
+            for (let i = 0; i < meshInstances.length; i++) {
+                if (!meshInstances[i].mesh.vertexBuffer.format.hasUv1) {
+                    hasUv1 = false;
+                    break;
+                }
+            }
+            if (hasUv1) {
+
```
Comment (willeastcott) [inline src/scene/lightmapper.js:319]:
```suggestion

            if (hasUv1) {
```

File: src/scene/lightmapper.js [hunk 14]
```diff
@@ -244,641 +491,526 @@ class Lightmapper {
      * Only lights with bakeDir=true will be used for generating the dominant light direction. Defaults to
      * pc.BAKE_COLORDIR.
      */
-    bake(nodes, mode) {
+    bake(nodes, mode = BAKE_COLORDIR) {
+
+        let device = this.device;
+        const startTime = now();
```
Comment (willeastcott) [inline src/scene/lightmapper.js:497]: Why was this moved outside the `// #ifdef PROFILER` block?

File: src/scene/lightmapper.js [hunk 15]
```diff
@@ -244,641 +491,526 @@ class Lightmapper {
...
+            this.scene._needsStaticPrepare = false;
+            this.revertStatic = true;
+        }
 
-        var renderOrModelMeshInstances = function (node) {
-            return node.render ? node.render.meshInstances : node.model.model.meshInstances;
-        };
+        // backup
+        this.fog = this.scene.fog;
+        this.ambientLight = this.scene.ambientLight.clone();
```
Comment (willeastcott) [inline src/scene/lightmapper.js:642]: Is this not allocated/defined in the Lightmapper constructor? Ideally, you'd create the `Color` there, and here, you'd just copy.

File: src/scene/lightmapper.js (hunk 8 above)
Comment (willeastcott) [inline src/scene/lightmapper.js:90]:
```suggestion
            this.lightBounds.center.copy(tempSphere.center);
            this.lightBounds.halfExtents.set(tempSphere.radius, tempSphere.radius, tempSphere.radius);
```

File: src/scene/lightmapper.js (hunk 6 above)
Comment (willeastcott) [reply src/scene/lightmapper.js:44]: Note that you'd also need this at the top:

```javascript
import { BoundingSphere } from '../shape/bounding-sphere.js';
```

**Review by willeastcott — APPROVED** [review]
Approved with a whole lot of quite minor comments.

File: src/scene/lightmapper.js (hunk 6 above)
Comment (mvaligursky) [reply src/scene/lightmapper.js:44]: done

File: examples/graphics/lights-baked-ao.html (hunk 4 above)
Comment (mvaligursky) [reply examples/graphics/lights-baked-ao.html:59]: done

File: src/scene/lightmapper.js (hunk 14 above)
Comment (mvaligursky) [reply src/scene/lightmapper.js:495]: we have a request to make performance stats available in all stats .. and considering this costs us nothing, it's a good start.

File: src/scene/lightmapper.js (hunk 15 above)
Comment (mvaligursky) [reply src/scene/lightmapper.js:640]: done, thanks!

File: src/scene/lightmapper.js [hunk 16]
```diff
@@ -297,7 +297,7 @@ class Lightmapper {
         }
 
         // mesh instances from render component
-        if (node.render && node.render.enabled) {
+        if (node.render?.enabled) {
```
Comment (Maksims) [inline src/scene/lightmapper.js:302]: Have "optional chaining" been investigated in terms of Babel. How does it compile into ES5. If it is good, please consider PR to update [CONTRIBUTING.md](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md).

Comment (mvaligursky) [reply src/scene/lightmapper.js:302]: this line
`if (node.render?.enabled)`

compiles to this
`if ((_node$render = node.render) != null && _node$render.enabled)`

Comment (Maksims) [reply src/scene/lightmapper.js:302]: And if you do 3-5 chains in single line, does it look ok?

Comment (mvaligursky) [reply src/scene/lightmapper.js:302]: Slightly longer example. The es5 is not the easiest to read unfortunately. 
From
```
        var graph;
        if (this.entity.model?.model?.graph) {
            graph = this.entity.model?.model?.graph;
        } else {
            // animating hierarchy without model
            graph = this.entity;
        }
```
to
```
var graph;

if ((_this$entity$model = this.entity.model) != null && (_this$entity$model$mo = _this$entity$model.model) != null && _this$entity$model$mo.graph) {
	var _this$entity$model2, _this$entity$model2$m;

	graph = (_this$entity$model2 = this.entity.model) == null ? void 0 : (_this$entity$model2$m = _this$entity$model2.model) == null ? void 0 : _this$entity$model2$m.graph;
} else {
	graph = this.entity;
}
```

Comment (Maksims) [reply src/scene/lightmapper.js:302]: Readability is not great indeed, but ES5 not meant to be read (unless debugging babel specific issues, which we haven't met yet). So probably is fine. Worth adding it to CONTRIBUTING.md.

Comment (willeastcott) [reply src/scene/lightmapper.js:302]: Done: https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md#whitelisted-es6-features

**Pull request merged**